### ISSUE TITLE
[Bug] Assigning an array only copies the first element

### ISSUE THREAD

**Issue by arcusmaximus** [WerWolv/ImHex#1413]
### Operating System

Windows

### What's the issue you encountered?

If you assign one array variable to another, only the first element gets copied; the remaining ones stay uninitialized.

### How can the issue be reproduced?

```
#include <std/io.pat>

u8 values[2] @ 0;

u8 gValues[2];
gValues = values;

std::print("{:02X} {:02X}", gValues[0], gValues[1]);
```

Expected: prints the first two bytes in the file.
What actually happens: prints the first byte followed by 00.

### ImHex Version

1.31.0 7d53636

### ImHex Build Type

- [X] Nightly or built from sources

### Installation type

Portable

**Comment by paxcut** [comment]
assigning arrays has never been supported or documented. You are better off placing a second array at same address as the first one
```cpp
#include <std/io.pat>

u8 values[2] @ 0;

u8 gValues[2] @ 0;

std::print("{:02X} {:02X}", gValues[0], gValues[1]);
```

**Comment by arcusmaximus** [comment]
The example is contrived to simplify things - I was trying this as a workaround for #1412, copying an array field to a global variable instead of passing it through a template argument.

I think assigning an array is such a basic feature that it shouldn't even need to be documented - and it certainly seems easier to implement than (checks recent commits) a graphing calculator?! 😅

**Comment by paxcut** [comment]
Assigning arrays like that is not defined in any 'normal' languages. Normally you loop over contents if you want a copy. I am also looking at the other bug but assigning arrays is not a well defined problem. You need to assume they are the same size or free and reallocate the size. Instead create a function that takes the array as a reference and then copy values.

**Comment by arcusmaximus** [comment]
I guess C# and Python aren't normal languages then, but fair enough, in C it's also not possible. Maybe ImHex could raise a syntax error like C then.

Copying item by item is indeed the next thing I thought of, but surprise surprise, that's also bugged:

```
#include <std/io.pat>

struct Wrapper
{
    u32 value;
};

Wrapper from[1];
from[0].value = 1;

Wrapper to[1];
to[0] = from[0];

std::print("{}", to[0].value);
```

This *is* something that would work in C, but in ImHex you get the following:

```
E: error[E0011]: Memory error.
E: Tried accessing out of bounds heap cell 2. This is a bug.
```

**Comment by paxcut** [comment]
```cpp
to[0] .value= from[0].value;
```

**Comment by arcusmaximus** [comment]
> to[0].value= from[0].value;

That's fine if you have one field, but not so much if you have 50. As before, I'm not looking for alternative solutions for these minimal examples - I'm pointing out basic language features that you'd expect to work but don't.

If even ImHex itself says "This is a bug", I hope we can agree there is in fact a bug here.

> there is no = operator defined for array elements that are udts.

If it's not supported and will never be supported, I'd expect the program to at least tell me so.

**Comment by paxcut** [comment]
I agree error messages could be better but it is not easy thing to do always.  you can copy udts as long as they are not array members. for example this works 
```cpp
#include <std/io.pat>

struct Wrapper
{
    u32 value;
};

Wrapper from @ 0;
Wrapper temp;

temp = from;
Wrapper to;
to = temp;

std::print("{}", to.value);
```

**Comment by paxcut** [comment]
it can also copy arrays i just update the last code i posted

**Comment by arcusmaximus** [comment]
>  you can copy udts as long as they are not array members

Sure. But this ticket is about arrays :)

> Arrays are hard because each element could be different in size and members

Even in C, we can solve this easily by populating the array with pointers instead.

> it can also copy arrays i just update the last code i posted. you need to wrap the array im a pattern

Great, but again - that's a workaround, not a fix.

**Comment by paxcut** [comment]
yes, i was just told that this is a bug that should be fixed

**Comment by arcusmaximus** [comment]
Here's another fun one (section included for easier reproducability, but happens with a regular file too):

```
#include <std/io.pat>
#include <std/mem.pat>

struct VertexAttribute
{
    u8 offset;
    u8 type;
};

struct VertexFormat
{
    VertexAttribute attributes[1];
};

struct Vertex
{
    std::print("{:02X}", parent.format.attributes[0].type);
    
    VertexAttribute attr = parent.format.attributes[0];
    std::print("{:02X}", attr.type);
};

struct Mesh
{
    VertexFormat format;
    Vertex vertex;
};

std::mem::Section section = std::mem::create_section("Test");
VertexAttribute attr @ 0 in section;
attr.offset = 0x01;
attr.type = 0x02;

Mesh mesh @ 0 in section;
```

Expectation: the type (0x02) gets printed twice.
Actual result: the `attr.type` variant prints the offset (0x01) instead. What 😄 

(And before you say "the first line works so just use that one" - imagine that it's not parent.format.attributes[0].type but parent.parent.parent.vertexFormat.attributes[gCurrentVertexAttributeIndex].type, and that you also want to fetch the offset and the vertex buffer index... You really don't want to repeat that whole mouthful each time.)
